### PR TITLE
Refactor dataset download process for efficiency, simplicity, and security

### PR DESCRIPTION
**Pull request by Nottlespike**
- Replace Git LFS dependency with custom Python script
- Utilize aria2c/wget for faster, multi-threaded downloads
- Implement data rank-based subset selection
- Streamline setup process and reduce dependencies
- Update README with new download instructions
- Fixed issue with ~/10B/H100.toml where it would attempt to access the root /data/ directory

This change improves download speeds, reduces system requirements,
and simplifies the dataset preparation workflow for users.

NOTE: The data rank-based subset selection logic should be verified
by maintainers to ensure it matches the intended implementation.

### CODE REVIEW

File: configs/10B/H100.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ z_loss = true
 
 [data]
 seq_length = 8192
-dataset_name_or_paths = "/data/datasets/fineweb-edu,/data/datasets/fineweb,/data/datasets/StackV1-popular,/data/datasets/dclm-baseline-1.0-parquet,/data/datasets/open-web-math"
```
Comment (samsja) [inline configs/10B/H100.toml:23]: we still want to keep the `/data/` as this is where we store our data for the 10b run

Comment (Nottlespike) [reply configs/10B/H100.toml:23]: I mean that works but is not ideal. You use the root /data directory? This requires you to run the script as sudo or be a root user. We can symlink to be backwards compatible but using the root /data is certainly a bad practice in my thoughts.

Comment (Nottlespike) [reply configs/10B/H100.toml:23]: Perhaps we should have a migration script? If we MUST keep it as is I can modify the PR to accomodate, but it would require the user to run the dataset script as sudo.

Comment (samsja) [reply configs/10B/H100.toml:23]: We are mainly using the 10b config for us right now, let's keep it like this. user can override it and we can fix it later

Comment (Nottlespike) [reply configs/10B/H100.toml:23]: > We are mainly using the 10b config for us right now, let's keep it like this. user can override it and we can fix it later

Yeah this would be a breaking change to the 10B run right now. Perhaps we could move this to a branch and merge it after the 10B is trained?

Comment (samsja) [reply configs/10B/H100.toml:23]: yeah could makes sense :smile:

File: scripts/subset_data.py [hunk 2]
```diff
@@ -18,56 +15,131 @@
...
+        
+        files = response.json()
+        for file in files:
+            if file['type'] == 'file':
+                all_files.append(file['path'])
+            elif file['type'] == 'directory':
+                traverse_files(file['path'])
+    
+    traverse_files()
+    return [f for f in all_files if f.endswith('.parquet')]
```
Comment (Jackmin801) [inline scripts/subset_data.py:47]: Nice solution. This gets every file in the dataset regardless of the dataset config though.
What do you think about using datasets BuilderConfig to retrieve datafiles? It would support dataset configs and be closest to how the HF dataset would parse the data files.

File: README.md [hunk 3]
```diff
@@ -122,18 +122,70 @@ export HF_HUB_ETAG_TIMEOUT=500
...
+Example for downloading files for data rank 5 in a training setup with data world size of 12, using the configuration from 10B/H100.toml:
+
 ```bash
-export GIT_LFS_SKIP_SMUDGE=1
-git clone git@hf.co:datasets/PrimeIntellect/fineweb-edu
-cd fineweb-edu/
-python3 ../subset_data.py --dataset_name PrimeIntellect/fineweb-edu --data_world_size 12 --data_rank 5
-bash rm-unused.sh
+python scripts/subset_data.py \
+  --dataset_name PrimeIntellect/fineweb-edu,PrimeIntellect/fineweb,PrimeIntellect/StackV1-popular,PrimeIntellect/dclm-baseline-1.0-parquet,PrimeIntellect/open-web-math \
```
Comment (Jackmin801) [inline README.md:151]: Lets change the dclm and openwebmath org to their respective orgs so `mlfoundations/dclm-baseline-1.0-parquet`, `open-web-math/open-web-math`.

We dont have dclm-baseline-1.0-parquet and open-web-math on PrimeIntellect atm because we didn't need to reshuffle them (they seemed shuffled enough, though maybe we could have reshuffled them to be sure and also to get nicer shard sizes).

**Review by Jackmin801 — APPROVED** [review]
Nice! Thanks for the PR :) Added some minor comments.
Ill give it a spin on my side to make sure its working and we should be able to merge without the changes to `H100.toml`

**Review by Jackmin801 — CHANGES_REQUESTED** [review]
Ah accidentally approved

File: README.md (hunk 3 above)
Comment (Nottlespike) [reply README.md:151]: Ah I see. I should have checked for all the respositories.

File: scripts/subset_data.py (hunk 2 above)
Comment (Nottlespike) [reply scripts/subset_data.py:47]: That would be a more robust solution that is closer to how the Hugging Face datasets library parses the data. It would require some small changes in data.py to handle this. I can modify the PR.

Comment (Nottlespike) [reply scripts/subset_data.py:47]: Ok this is tougher than expected but necessary. Will likely have to rewrite to hande dataset configs.
